### DISCUSSION TITLE
Array Reduction

### DISCUSSION THREAD

**Discussion by goerlibe** [discopop-project/discopop#357] [Development]
We currently struggle with reductions into array variables. For some examples a false doall pattern is reported. For others, no pattern is reported.

The OpenMP Microbench currently also does not support measuring reduction overhead for array variables.

**Comment by goerlibe** [comment]
**Bug**:
DiscoPoP suggests doall patterns for both the outer and inner loop of the following example.
**Expected**:
DiscoPoP suggests reductions (or at least not a false doall)

**Variation**: changing the indices in `sums[j] += matrix[j*N + i];` to `sums[i] += matrix[i*N + j];`leads to no pattern being suggested. A reduction would be possible.

**Additional Information**:
- Currently array reductions are disabled in discopop_explorer, but with the new improved dependencies we should be able to get them to work with reasonable effort.
- It seems like the detection of the reduction (reductions.txt) is wrong.

**Example**:
```c
#include <stdlib.h>
#include <stdio.h>
​
​
#define N 5
​
int main() {
    // loop indices
    int i,j;
​
    // initialize matrix
    double matrix[N*N];
    for (i = 0; i < N*N; i++)
    {
        matrix[i] = (double) i;
    }
    
    // calculate sums per column
    double sums[N] = {0}; 
    // # pragma omp parallel for private(i,j) shared(matrix) reduction(+:sums)
    for (i=0; i<N; i++)
    {
        for(int j=0; j<N; j++) {
            sums[j] += matrix[j*N + i];
        }
    }
​
    // print result
    for (i = 0; i < N; i++)
    {
        printf("%.0f ", sums[i]);
    }
    printf("\n");
​
    return 0;
}
```

**Reply by goerlibe** [reply]
Can confirm: injecting a corrected reductions.txt makes it work as expected. So the problem lies only in detecting the reduction operations.

**Reply by lukasrothenberger** [reply]
okay, good. That significantly narrows down the search space.
